presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java

```diff
@@ -253,14 +279,105 @@ public Partitioning translateVariable(Function<VariableReferenceExpression, Vari
+            RowExpression rightArgument = other.arguments.get(i);
+            if (leftArgument instanceof ConstantExpression) {
+                arguments.add(leftArgument);
+            }
+            else if (rightArgument instanceof ConstantExpression) {
+                arguments.add(rightArgument);
+            }
+            else if (leftArgument instanceof VariableReferenceExpression && rightArgument instanceof VariableReferenceExpression) {
+                VariableReferenceExpression leftVariable = (VariableReferenceExpression) leftArgument;
+                VariableReferenceExpression rightVariable = (VariableReferenceExpression) rightArgument;
+                checkArgument(leftVariable.getType().equals(rightVariable.getType()), "incompatible types: %s != %s", leftVariable.getType(), rightVariable.getType());
+                arguments.add(new SpecialFormExpression(COALESCE, leftVariable.getType(), ImmutableList.of(leftVariable, rightVariable)));
+            }
+            else {
+                return Optional.empty();
+            }
+        }
+        return Optional.of(new Partitioning(this.handle, arguments.build()));
+    }
+
+    public Optional<Partitioning> translateRowExpression(Map<VariableReferenceExpression, RowExpression> inputToOutputMappings, Map<VariableReferenceExpression, RowExpression> assignments, TypeProvider types)
+    {
+        ImmutableList.Builder<RowExpression> newArguments = ImmutableList.builder();
+        for (RowExpression argument : arguments) {
+            if (argument instanceof ConstantExpression) {
+                newArguments.add(argument);
+            }
+            else if (argument instanceof VariableReferenceExpression) {
+                if (!inputToOutputMappings.containsKey(argument)) {
+                    return Optional.empty();
+                }
+                newArguments.add(inputToOutputMappings.get(argument));
+            }
+            else {
+                checkArgument(argument instanceof SpecialFormExpression && ((SpecialFormExpression) argument).getForm().equals(COALESCE), "Expect argument to be COALESCE");
+                Set<RowExpression> coalesceArguments = ImmutableSet.copyOf(((SpecialFormExpression) argument).getArguments());
+                checkArgument(coalesceArguments.stream().allMatch(VariableReferenceExpression.class::isInstance), "Expect arguments of COALESCE to be VariableReferenceExpression");
+                // We are using the property that the result of coalesce from full outer join keys would not be null despite of the order
+                // of the arguments. Thus we extract and compare the variables of the COALESCE as a set rather than compare COALESCE directly.
+                VariableReferenceExpression translated = null;
+                for (Map.Entry<VariableReferenceExpression, RowExpression> entry : assignments.entrySet()) {
+                    if (isExpression(entry.getValue())) {
```

Comment:
question: Is it still possible to have AST Expressions in the plan?

Reply:
Depending on the progress of #12828. Currently `AddExchanges` happens before `TranslateExpressions` which means this will all be AST expressions. I'm adding handling to both but I can remove the real RowExpression one I think. Once translation happens before `AddExchanges` we can remove the AST part. Let me know what you prefer.